### ISSUE TITLE
Jsonschema Forward compatibility fails when deleting an optional field

### ISSUE THREAD

**Issue by adiohana** [Apicurio/apicurio-registry#1710]
According to [this](https://docs.confluent.io/platform/current/schema-registry/avro.html#), deleting an optional field is allowed in Forward compatibility.

Having this schema:

```json
{
  "title": "Person",
  "type": "object",
  "properties": {
    "firstName": {
      "type": "string",
      "description": "The person's first name."
    },
    "lastName": {
      "type": "string",
      "description": "The person's last name."
    },
    "age": {
      "description": "Age in years which must be equal to or greater than zero.",
      "type": "integer",
      "minimum": 0
    }
  }
}
```

and setting compatibility rule to Forward, and trying to add a new version:

```json
{
  "title": "Person",
  "type": "object",
  "properties": {
    "firstName": {
      "type": "string",
      "description": "The person's first name."
    },
    "age": {
      "description": "Age in years which must be equal to or greater than zero.",
      "type": "integer",
      "minimum": 0
    }
  }
}
```

the compatibility check is failing with error:

```
{causes: [{description: "OBJECT_TYPE_PROPERTY_SCHEMAS_NARROWED", context: "/propertySchemasAdded"}],…}
causes: [{description: "OBJECT_TYPE_PROPERTY_SCHEMAS_NARROWED", context: "/propertySchemasAdded"}]
detail: "io.apicurio.registry.rules.RuleViolationException
error_code: 409
message: "Incompatible artifact: person [JSON], num of incompatible diffs: {1}"
name: "RuleViolationException"
```

**Comment by bobmorley** [comment]
Would like to increase priority on this -- the `Forward` / `Forward Transitive` do not work if you add an optional field.

**Version 1**
```
{
"$schema":"http://json-schema.org/draft-07/schema",
"$id":"https://example.auvik.com/goka/payload.json",
"title":"Square",
"description":"The schema for a unit-less square.",
"type":"object",
"properties":{
	"length":{
		"type":"integer",
		"description":"Length of one side of the square"
	}
},
"additionalProperties":false
}
```

**Version 2**
```
{
		"$schema":"http://json-schema.org/draft-07/schema",
		"$id":"https://example.auvik.com/goka/payload.json",
		"title":"Rectangle",
		"description":"The schema for a unit-less Rectangle that is compatible with the original square schema.",
		"type":"object",
		"properties":{
			"length":{
				"type":"integer",
				"description":"Length of one side of the square"
			},
			"width":{
				"type":"integer",
				"description":"Width of the rectangle"
			}
		},
		"additionalProperties":false
}
```

**Failure**
`2022-04-19 19:00:47 INFO <_> [io.apicurio.common.apps.logging.audit.AuditLogService] (executor-thread-276) apicurio.audit action="createArtifactVersion" result="failure" src_ip="172.17.0.1" error_msg="Incompatible artifact: integration.test.schema.id [JSON], num of incompatible diffs: {1}, list of diff types: [OBJECT_TYPE_PROPERTY_SCHEMAS_NARROWED at /propertySchemasRemoved]" update_state="1.1.0" artifact_id="integration.test.schema.id"`

 If I change my global rules to `Backward` then creating and adding the second schema to the artifact works without issue.

**Comment by jsenko** [comment]
I don't think adding an optional field is in general backwards compatible (and the reverse, removing is forward compatible). Consider this example:
Original
```
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
    "properties": {
      "name": {
        "type": "string"
      }
  }
}
```
matches
```
{
  "name": "Foo"
}
```
but also matches
```
{
  "name": "Foo",
  "id": "652d88de-876d-4bae-88f3-c477cf0febf2"
}
```
When you add an optional field that matches an existing property, but with a different type like this:
```
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
    "properties": {
      "name": {
        "type": "string"
      },
      "id": {
        "type": "number"
      }
  }
}
```
The first match is still valid, but the second match is not, which breaks backwards compatibility. I realize that this is an edge case, but the compatibility checker has to be general and conservative in this way. We've been thinking about a possibility to configure it to be less strict. I'm interested in your ideas, WDYT?

**Comment by jsenko** [comment]
The reverse case, removing an optional field should be backwards compatible.

**Comment by jsenko** [comment]
I'm thinking however that my argument may not be valid if there is `"additionalProperties":false` set.

**Comment by jsenko** [comment]
Related information:
- https://github.com/confluentinc/schema-registry/issues/2121
- https://yokota.blog/2021/03/29/understanding-json-schema-compatibility/

**Comment by bobmorley** [comment]
I am unclear why this issue was closed. Multiple sources indicate that to be backward/forward compatible you may add/remove optional fields. Are we saying that is not true? Or only true if `additionalProperties: false` is set?

Our intent was to use `FULL` for validation, but we have had to use `BACK` because of what i perceive as this being a bug.

**Comment by EricWittmann** [comment]
I think Jakub's argument is that your expectation about adding an optional field being backwards compatible is only true when `additionalProperties` is `false`.  Otherwise you run into the edge case issue he outlined above.

And also I think he's saying that if you DO have `additionalProperties` set to `false` then our compatibility checker actually works as you are expecting.

That said, he also indicated that we're open to making the compatibility checking less strict or at least configurable.

Thoughts?

**Comment by jsenko** [comment]
Yes, the compatibility checker must be *pessimistic* because it works based on general rules. There was already some interest from the community to introduce a less strict checking, but I'm not yet sure how would it look like in practice. 
A straightforward alternative would be to add an option to force an override, on user's own risk if they are 100% sure that it's safe for their application. Registry would then need to take this into account when doing transitive checks.

**Comment by big-andy-coates** [comment]
I ran into the same issue. In fact, if you're looking for full compatibility then you're out of luck. While it's technically possible to add and remove optional properties with full compatibility checks, there are some serious hoops to jump through. This makes it, IMHO, not workable as a solution.

I've detailed my findings in a [quick post](https://www.creekservice.org/articles/2024/01/08/json-schema-evolution-part-1.html), the second part of which proposes a more user friendly approach to implementing JSON Schema compatibility checks.

I've raised https://github.com/confluentinc/schema-registry/issues/2927 to get feedback and look at the potential of including this in the Schema Registry.
